File: src/tacky.fs
module Tacky

type TackyUnaryOperator = Complement | Negate | Not

type TackyBinaryOperator =
    | Add
    | Subtract
    | Multiply
    | Divide
    | Mod
    | Equal
    | NotEqual
    | LessThan
    | LessOrEqual
    | GreaterThan
    | GreaterOrEqual

// we need a custom comparison function for constants to make sure that 0.0 and
// -0.0 don't compare equal
let const_compare a b =
    match (a, b) with
    | Const.ConstDouble d1, Const.ConstDouble d2 when d1 = d2 ->
        compare (System.Math.CopySign(1.0, d1)) (System.Math.CopySign(1.0, d2))
    | _ -> compare a b

type TackyVal =
    | Constant of Const.t
    | Var of string

let show_tacky_val = function Constant c -> Const.show c | Var v -> v
let pp_tacky_val (fmt: System.IO.TextWriter) v = fmt.Write(show_tacky_val v)

// TODO maybe this should be in a separate module?
let type_of_val = function
    // note: this reports the type of ConstChar as SChar instead of Char, doesn't
    // matter in this context
    | Constant c -> Const.type_of_const c
    | Var v -> (Symbols.get v).symType

type TackySrcDst = { src: TackyVal; dst: TackyVal }

type TackyUnaryInfo = { op: TackyUnaryOperator; src: TackyVal; dst: TackyVal }

type TackyBinaryInfo = {
    op: TackyBinaryOperator
    src1: TackyVal
    src2: TackyVal
    dst: TackyVal
}

type TackyAddPtrInfo = {
    ptr: TackyVal
    index: TackyVal
    scale: int
    dst: TackyVal
}

type TackyCopyToOffsetInfo = { src: TackyVal; dst: string; offset: int }

type TackyCopyFromOffsetInfo = { src: string; offset: int; dst: TackyVal }

type TackyFunCallInfo = { f: string; args: TackyVal list; dst: TackyVal option }

type TackyInstruction =
    | Return of TackyVal option
    | SignExtend of TackySrcDst
    | ZeroExtend of TackySrcDst
    | DoubleToInt of TackySrcDst
    | IntToDouble of TackySrcDst
    | DoubleToUInt of TackySrcDst
    | UIntToDouble of TackySrcDst
    | Truncate of TackySrcDst
    | Unary of TackyUnaryInfo
    | Binary of TackyBinaryInfo
    | Copy of TackySrcDst
    | GetAddress of TackySrcDst
    | Load of {| src_ptr: TackyVal; dst: TackyVal |}
    | Store of {| src: TackyVal; dst_ptr: TackyVal |}
    | AddPtr of TackyAddPtrInfo
    | CopyToOffset of TackyCopyToOffsetInfo
    | CopyFromOffset of TackyCopyFromOffsetInfo
    | Jump of string
    | JumpIfZero of TackyVal * string
    | JumpIfNotZero of TackyVal * string
    | Label of string
    | FunCall of TackyFunCallInfo

type TackyFunctionDef = {
    name: string
    ``global``: bool
    ``params``: string list
    body: TackyInstruction list
}

type TackyStaticVariableDef = {
    name: string
    t: Types.t
    ``global``: bool
    init: Initializers.static_init list
}

type TackyStaticConstantDef = {
    name: string
    t: Types.t
    init: Initializers.static_init
}

type TackyTopLevel =
    | Function of TackyFunctionDef
    | StaticVariable of TackyStaticVariableDef
    | StaticConstant of TackyStaticConstantDef

type TackyProgram = Program of TackyTopLevel list

File: src/Tacky.rs
//! Three-address ("TACKY") intermediate representation.

use super::{
    Const::{self, ConstValue},
    Initializers::StaticInit,
    Symbols,
    Types::Type,
};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TackyUnaryOperator {
    Complement,
    Negate,
    Not,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TackyBinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Mod,
    Equal,
    NotEqual,
    LessThan,
    LessOrEqual,
    GreaterThan,
    GreaterOrEqual,
}

#[derive(Clone, Debug, PartialEq)]
pub enum TackyVal {
    Constant(ConstValue),
    Var(String),
}

pub fn typeOfVal(value: &TackyVal) -> Type {
    match value {
        TackyVal::Constant(value) => Const::type_of_const(*value),
        TackyVal::Var(name) => Symbols::get(name).symType,
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct TackySrcDst {
    pub src: TackyVal,
    pub dst: TackyVal,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyUnaryInfo {
    pub op: TackyUnaryOperator,
    pub src: TackyVal,
    pub dst: TackyVal,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyBinaryInfo {
    pub op: TackyBinaryOperator,
    pub src1: TackyVal,
    pub src2: TackyVal,
    pub dst: TackyVal,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyAddPtrInfo {
    pub ptr: TackyVal,
    pub index: TackyVal,
    pub scale: i32,
    pub dst: TackyVal,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyCopyToOffsetInfo {
    pub src: TackyVal,
    pub dst: String,
    pub offset: i32,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyCopyFromOffsetInfo {
    pub src: String,
    pub offset: i32,
    pub dst: TackyVal,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyFunCallInfo {
    pub f: String,
    pub args: Vec<TackyVal>,
    pub dst: Option<TackyVal>,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyLoadInfo {
    pub src_ptr: TackyVal,
    pub dst: TackyVal,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyStoreInfo {
    pub src: TackyVal,
    pub dst_ptr: TackyVal,
}

#[derive(Clone, Debug, PartialEq)]
pub enum TackyInstruction {
    Return(Option<TackyVal>),
    SignExtend(TackySrcDst),
    ZeroExtend(TackySrcDst),
    DoubleToInt(TackySrcDst),
    IntToDouble(TackySrcDst),
    DoubleToUInt(TackySrcDst),
    UIntToDouble(TackySrcDst),
    Truncate(TackySrcDst),
    Unary(TackyUnaryInfo),
    Binary(TackyBinaryInfo),
    Copy(TackySrcDst),
    GetAddress(TackySrcDst),
    Load(TackyLoadInfo),
    Store(TackyStoreInfo),
    AddPtr(TackyAddPtrInfo),
    CopyToOffset(TackyCopyToOffsetInfo),
    CopyFromOffset(TackyCopyFromOffsetInfo),
    Jump(String),
    JumpIfZero(TackyVal, String),
    JumpIfNotZero(TackyVal, String),
    Label(String),
    FunCall(TackyFunCallInfo),
}

#[derive(Clone, Debug, PartialEq)]
pub struct TackyFunctionDef {
    pub name: String,
    pub global: bool,
    pub params: Vec<String>,
    pub body: Vec<TackyInstruction>,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyStaticVariableDef {
    pub name: String,
    pub t: Type,
    pub global: bool,
    pub init: Vec<StaticInit>,
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyStaticConstantDef {
    pub name: String,
    pub t: Type,
    pub init: StaticInit,
}
#[derive(Clone, Debug, PartialEq)]
pub enum TackyTopLevel {
    Function(TackyFunctionDef),
    StaticVariable(TackyStaticVariableDef),
    StaticConstant(TackyStaticConstantDef),
}
#[derive(Clone, Debug, PartialEq)]
pub struct TackyProgram(pub Vec<TackyTopLevel>);
